app/parse.py

```diff
@@ -13,10 +22,108 @@ class Product:
     price: float
     rating: int
     num_of_reviews: int
+    additional_info: dict
+
+PRODUCT_FIELDS = [field.name for field in fields(Product)]
+
+_driver: WebDriver | None = None
+
+def get_driver() -> WebDriver:
+    return _driver
+
+
+def set_driver(new_driver: WebDriver) -> None:
+    global _driver
+    _driver = new_driver
+
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)8s: %(message)s]",
+    handlers=[
+        logging.FileHandler("parser.log"),
+        logging.StreamHandler(sys.stdout)
+    ]
+)
+
+
+def parse_hdd_block_proces(product_soup: BeautifulSoup) -> dict[str, float]:
+    detail_url = urljoin(BASE_URL, product_soup.select_one(".title")["href"])
+    driver = get_driver()
+    driver.get(detail_url)
+    swatches = driver.find_element(By.CLASS_NAME, "swatches")
+    buttons = swatches.find_elements(By.TAG_NAME, "button")
+
+    prices = {}
+
+    for button in buttons:
+        if not button.get_property("disabled"):
+            button.click()
+            prices[button.get_property("value")] = float(driver.find_element(
+                By.CLASS_NAME, "price"
+            ).text.replace("$", ""))
+
+    return prices
+
+
+def parse_single_product(product_soup: BeautifulSoup) -> Product:
+    hdd_prices = parse_hdd_block_proces(product_soup)
+    return Product(
+        title=product_soup.select_one(".title")["title"],
+        description=product_soup.select_one(".description").text,
+        price=float(product_soup.select_one(".price").text.replace("$", "")),
+        rating=int(product_soup.select_one("p[data-rating]")["data-rating"]),
+        num_of_reviews=int(
+            product_soup.select_one(".ratings > p.review-count").text.split()[0]
+        ),
+        additional_info={"hdd_prices": hdd_prices}
+    )
+
+
+def get_num_pages(page_soup: BeautifulSoup) -> int:
+    pagination = page_soup.select_one(".pagination")
+
+    if pagination is None:
+        return 1
+
+    return int(pagination.select("li")[-2].text)
+
+
+def get_single_page_products(page_soup: BeautifulSoup) -> [Product]:
+    products = page_soup.select(".thumbnail")
+
+    return [parse_single_product(product_soup) for product_soup in products]
+
+
+def get_laptop_products() -> list[Product]:
+    page = requests.get(HOME_URL).content
+    first_page_soup = BeautifulSoup(page, "html.parser")
+
+    num_pages = get_num_pages(first_page_soup)
+
+    all_products = get_single_page_products(first_page_soup)
+
+    for page_num in range(2, num_pages + 1):
+        page = requests.get(HOME_URL, {"page": page_num}).content
+        soup = BeautifulSoup(page, "html.parser")
+        all_products.extend(get_single_page_products(soup))
+        break
```

Comment:
The `break` statement here will cause the loop to exit after the first iteration, which means it will not process all pages. If the intention is to iterate through all pages, this `break` statement should be removed.